src/ros_filter.cpp

```diff
@@ -2747,6 +2798,87 @@ namespace RobotLocalization
 
     return canTransform;
   }
+
+  template<typename T>
+  void RosFilter<T>::saveFilterState(FilterBase& filter)
+  {
+    FilterState state;
+    state.state_ = filter.getState();
+    state.estimate_error_covariance_ = filter.getEstimateErrorCovariance();
+    state.time_ = filter.getLastMeasurementTime();
+    filterStateHistory_.push_back(state);
+    ros::Time ts;
+    ts.fromSec(state.time_);
+    RF_DEBUG("\n---- RosFilter::saveFilterState for timestamp " << ts << " size = " << filterStateHistory_.size() << "----\n");
+  }
+
+  template<typename T>
+  bool RosFilter<T>::restoreFilterAndMeasurementsBefore(const double t)
+  {
+    RF_DEBUG("\n----- RosFilter::restoreFilterAndMeasurementsBefore (" << t << ") -----\n");
+    // Remove all filter states older than t. This yields either the latest
+    // filter state before t or an empty queue. In that case the smoothing
+    // window is too small and we cannot reset the filter.
+    //
+
+    int num_popped_filter_updates = 0;
+    while (!filterStateHistory_.empty()
+        && filterStateHistory_.back().time_ > t)
+    {
+      filterStateHistory_.pop_back();
+      num_popped_filter_updates++;
+    }
+    RF_DEBUG("\n---- Popped " << num_popped_filter_updates << " filter updates. Queue is empty/size? " << filterStateHistory_.empty() << "/" << filterStateHistory_.size() << "\n");
+
+    if (filterStateHistory_.empty()) return false;
+
+    // Reset filter to the latest state from the queue.
+    const FilterState &old = filterStateHistory_.back();
+    filter_.setState(old.state_);
+    filter_.setEstimateErrorCovariance(old.estimate_error_covariance_);
+    filter_.setLastMeasurementTime(old.time_);
+
+    // Find all measurements from the history whose timestamp is > t
+    // and push them to the measurements queue for integration.
+    // This re-establishes our invariant: measurementHistory_.back().time_ < measurementQueue_.top()
+    while (!measurementHistory_.empty()
+           && measurementHistory_.back().time_ > t)
+    {
+      measurementQueue_.push(measurementHistory_.back());
+      measurementHistory_.pop_back();
```

Comment:
Same thing: we're destroying the measurement history, which means this will only work for a single measurement. If another measurement needs to take advantage of the history, it won't be able to. Correct?